### DISCUSSION TITLE
CLI command for assembling (but not training) a model

### DISCUSSION THREAD

**Discussion by kinghuang** [explosion/spaCy#7688] [New Features & Project Ideas]
We've recently upgraded to spaCy 3, thanks to the commercial support! We do a "two-phase" approach to building our language model, where the `ner` component is trained and versioned in one project, and then assembled with other components in another project.

The final assembly is being done by running [spacy train …](https://spacy.io/api/cli#train) with a [training config](https://spacy.io/api/data-formats#config) that has every trainable pipeline component frozen. The goal of the assembly isn't to train any of the components, but to assemble a bunch of components together into a complete NLP model/package. However, using the `train` command requires train and dev data to be passed in and causes a training pipeline to run, even though there's no actual training being performed.

It would be useful if there was a command akin to `assemble`, that does all the package assembly that the `train` command does (e.g., collecting components and generating a nlp module), without invoking any training.

## Your Environment

- Operating System: Debian 10 (python container)
- Python Version Used: 3.8.
- spaCy Version Used: 3.0.5
- Environment Information: N/A

**Comment by polm** [comment]
Thanks for the suggestion, that's an interesting use case. Can you clarify why you assemble the model in a separate step rather than loading all the components, even if frozen, in the initial training step?

As far as the changes necessary for that, if you look at [the implementation of the train command](https://github.com/explosion/spaCy/blob/de4f4c9b8a46395b5f2be5fa80afd746c4a7f3cb/spacy/cli/train.py#L59), it looks like you could just skip running the last line and replace it with saving the model.

**Reply by polm** [reply]
Note that [this description](https://github.com/explosion/spaCy/discussions/7498#discussioncomment-512402) of building a pipeline in code may also be a useful reference - sticking trained components into a pipeline and serializing them is actually quite simple.

**Reply by kinghuang** [reply]
Separation of concerns. The two major parts of the final model are the NER component and a bunch of custom components. We are building and versioning the trained NER component as one asset, which is then combined with the components to produce a final versioned model.

I have looked at the underlying code, and that's exactly what I mean. I'm basically looking for a CLI command that's the same as `train_cli()`, but just outputs the assembled model instead of calling `train()`.

**Reply by kinghuang** [reply]
We used to have custom scripts for assembling the model, but the training config file is way more flexible. I just don't want to actually *train* in some cases.

**Reply by svlandeg** [reply]
Hi King,

I actually had this written down as a task on my TODO list ;-)

I think some type of "assemble" command would make a lot of sense to ensure you can keep using the config-based format even for steps where you're not actually training. So yes I think this could be a useful feature in many use-cases! (or you can do the "train for 1 epoch trick ;-))

**Reply by kinghuang** [reply]
Haha, it took me a little while to realize what was happening there. That “train for 1 epoch trick” is pretty handy workaround right now! 😄

**Comment by svlandeg** [comment]
PR with the proposal for implementation: https://github.com/explosion/spaCy/pull/7783
This will be available from 3.0.6 onwards!
